### PR TITLE
feat(core): Add Google Cloud Logging JSON formatter

### PR DESCRIPTION
**Pull request by DRACULA1729**
## Description

Meltano's `json_formatter` emits log lines using the structlog keys `event`, `level`, and `timestamp`. Google Cloud Logging (formerly Stackdriver) recognizes a JSON line as a structured entry only when it contains the special fields `severity`, `message`, and `timestamp`. As a result, logs produced by `json_formatter` were ingested as plain text at `INFO` severity, and severity-based filtering and alerting were unavailable. This is the behavior reported in #9420.

This PR adds `google_cloud_logging_formatter`, which:

- renames `event` to `message` and `level` to `severity` (upper-cased, so it matches the Google Cloud Logging severities), and
- emits a `logging.googleapis.com/sourceLocation` field (file, line, function) when `callsite_parameters` is enabled.

With this formatter, logs are ingested as structured entries with the correct severity, without requiring a downstream agent (e.g. the Ops Agent) to remap fields.

Example output:

```json
{"timestamp": "2025-07-29T09:25:19Z", "severity": "WARNING", "message": "There was an error in the application.", "logging.googleapis.com/sourceLocation": {"file": "/app/run.py", "line": "42", "function": "main"}}
```

Changes:

- `src/meltano/core/logging/formatters.py`: add `google_cloud_logging_formatter` and its processor.
- `src/meltano/core/logging/__init__.py`: export the new formatter.
- `tests/meltano/core/logging/test_formatters.py`: unit tests for severity/message mapping, all severity levels, source location, and exceptions.
- `docs/docs/guide/logging.md`: update the Google Cloud logging section to use the new formatter.

## Checklist

- [x] I have read the [contribution guide](https://docs.meltano.com/contribute/merge/)

### Was generative AI tooling used to co-author this PR?

- [x] Yes (Claude Code)

Generated-by: Claude Code following [the agentic coding guidelines](https://github.com/meltano/meltano/blob/main/CONTRIBUTING.md#agentic-coding)

## Related Issues

* Closes #9420

## Summary by Sourcery

Add a Google Cloud Logging–compatible JSON formatter and document its usage for structured logging on GCP.

New Features:
- Introduce google_cloud_logging_formatter to emit JSON logs using Google Cloud Logging special fields for severity, message, and timestamp.

Enhancements:
- Populate Google Cloud Logging sourceLocation metadata from callsite parameters when enabled.

Documentation:
- Update logging guide to describe Google Cloud Logging behavior and show configuration examples using google_cloud_logging_formatter.

Tests:
- Add unit tests covering Google Cloud Logging formatter severity mapping, source location emission, and exception handling.

### CODE REVIEW

**Comment by DRACULA1729** [comment]
Thanks for the review. All three points are addressed in 6c8e1ebb0:

- `_google_cloud_logging_processor` now coerces the level with `str(level).upper()`, so a non-string level can't break it.
- Added edge-case tests for partial callsite parameters and for the case where no `sourceLocation` is emitted.
- Added a `show_locals` test that mirrors `test_json_formatter_locals`.

**Comment by DRACULA1729** [comment]
@sourcery-ai resolve

**Comment by DRACULA1729** [comment]
Hey @edgarrmondragon, when you get a chance could you take a look at this one? It's the code fix for #9420 (GCP logging severity mapping), and CI is green. Happy to adjust anything. Thanks!

File: src/meltano/core/logging/formatters.py [hunk 1]
```diff
@@ -263,6 +263,98 @@ def json_formatter(
...
+        *_processors_from_kwargs(
+            callsite_parameters=callsite_parameters,
+            dict_tracebacks=dict_tracebacks,
+            show_locals=show_locals,
+        ),
+        structlog.stdlib.ProcessorFormatter.remove_processors_meta,
+        structlog.stdlib.ExtraAdder(),
+        _google_cloud_logging_processor,
+        structlog.processors.JSONRenderer(),
+    )
```
Comment (edgarrmondragon) [inline src/meltano/core/logging/formatters.py:337]: This is almost identical to the `json_formatter` implementation. What do you think about adding a new flag to that formatter so that it emits GCL-compatible logs?

e.g.

```yaml
version: 1
disable_existing_loggers: false

formatters:
  google_cloud:
    (): meltano.core.logging.json_formatter
    profile: google-cloud-logging
```

I'm happy to "bikeshed" a bit on the exact keyword and option name, but you get the idea.

**Review by edgarrmondragon — COMMENTED** [review]
Thanks @DRACULA1729.

I have a question about your approach vs expanding options of an existing logger.

File: src/meltano/core/logging/formatters.py (hunk 1 above)
Comment (DRACULA1729) [reply src/meltano/core/logging/formatters.py:337]: Yeah, agreed, no point carrying a near-duplicate of `json_formatter`.

I'm thinking a `profile` option on `json_formatter` that takes a string, defaulting to `None` so nothing changes for existing configs. I'd lean string over a `google_cloud_logging=True` boolean so we're not adding a new flag every time another preset comes up:

```yaml
formatters:
  google_cloud:
    (): meltano.core.logging.json_formatter
    profile: google-cloud-logging
```

Under the hood it just tacks on the existing GCL processor when `profile="google-cloud-logging"`. Does `profile` work for you, or would you rather call it something like `preset`? Once you're happy with the name I'll drop the standalone `google_cloud_logging_formatter`, fold it into `json_formatter`, and update the docs and tests.

Comment (edgarrmondragon) [reply src/meltano/core/logging/formatters.py:337]: I do prefer `preset`. Go for it!

Comment (DRACULA1729) [reply src/meltano/core/logging/formatters.py:337]: Done in 39aae60c8. Went with `preset` like you preferred, so `json_formatter` now takes `preset="google-cloud-logging"` and tacks on the GCL processor. The standalone formatter is gone, an unknown preset raises `ValueError`, and the default (`preset=None`) output is unchanged. Docs and tests updated to match.

File: src/meltano/core/logging/formatters.py [hunk 2]
```diff
@@ -259,10 +277,70 @@ def json_formatter(
...
         structlog.stdlib.ProcessorFormatter.remove_processors_meta,
         structlog.stdlib.ExtraAdder(),
+        *preset_processors,
         structlog.processors.JSONRenderer(),
     )
 
 
+# Value accepted by the json_formatter `preset` option for Google Cloud Logging.
+GOOGLE_CLOUD_LOGGING_PRESET = "google-cloud-logging"
+
```
Comment (edgarrmondragon) [inline src/meltano/core/logging/formatters.py]: Should we move this further up the file, since it's used in `json_formatter`?

**Review by edgarrmondragon — COMMENTED** [review]
Thanks for following through with the changes @DRACULA1729. Just one small nit.

File: src/meltano/core/logging/formatters.py (hunk 2 above)
Comment (DRACULA1729) [reply src/meltano/core/logging/formatters.py]: Good call, moved it up in 84145a655 so the preset constant and processor are defined before `json_formatter`.

**Comment by DRACULA1729** [comment]
Heads up @edgarrmondragon, I just force-pushed to fix the author email on a few of my commits, they'd gone in with the wrong one. It only touched commit metadata, the tree is identical and your commits are unchanged apart from new SHAs. If you've got a local copy of the branch, you'll want to re-fetch before pushing again. Sorry for the noise!

**Comment by edgarrmondragon** [comment]
> Heads up @edgarrmondragon, I just force-pushed to fix the author email on a few of my commits, they'd gone in with the wrong one. It only touched commit metadata, the tree is identical and your commits are unchanged apart from new SHAs. If you've got a local copy of the branch, you'll want to re-fetch before pushing again. Sorry for the noise!

Thanks for the heads up @DRACULA1729!

**Review by edgarrmondragon — APPROVED** [review]
LGTM!

**Comment by edgarrmondragon** [comment]
@all-contributors please add @DRACULA1729 for code
